makeflow/src/makeflow_module_resource_monitor.c

```diff
@@ -0,0 +1,345 @@
+		debug(D_MAKEFLOW_HOOK, "Wrapper written to %s", df->filename);
+		makeflow_log_file_state_change(n->d, df, DAG_FILE_STATE_EXISTS);
+	} else {
+		debug(D_MAKEFLOW_HOOK, "Failed to create wrapper: errno %d, %s", errno, strerror(errno));
+		return MAKEFLOW_HOOK_FAILURE;
+	}
+	free(cmd);
+
+	return MAKEFLOW_HOOK_SUCCESS;
+}
+
+int makeflow_monitor_move_output_if_needed(struct dag_node *n, struct batch_queue *queue)
+{
+	if (!batch_queue_supports_feature(queue, "output_directories")) {
+		char *log_prefix = set_log_prefix(n);
+		char *output_prefix = xxstrdup(path_basename(log_prefix));
+
+		if (!strcmp(log_prefix, output_prefix)) { // They are in the same location so no move
+			free(log_prefix);
+			free(output_prefix);
+			return MAKEFLOW_HOOK_SUCCESS;
+		}
+
+		char *old_path = string_format("%s.summary", output_prefix);
+		char *new_path = string_format("%s.summary", log_prefix);
+		if (rename(old_path, new_path) == -1) {
+			debug(D_ERROR|D_MAKEFLOW_HOOK, "Error moving Resource Monitor output %s:%s. %s\n", old_path, new_path,
+					strerror(errno));
+			return MAKEFLOW_HOOK_FAILURE;
+		}
+		free(old_path);
+		free(new_path);
+
+		if (monitor->enable_time_series) {
+			char *old_path = string_format("%s.series", output_prefix);
+			char *new_path = string_format("%s.series", log_prefix);
+			if (rename(old_path, new_path) == -1) {
+				debug(D_ERROR|D_MAKEFLOW_HOOK, "Error moving Resource Monitor output %s:%s. %s\n", old_path,
+						new_path, strerror(errno));
+				return MAKEFLOW_HOOK_FAILURE;
+			}
+			free(old_path);
+			free(new_path);
+		}
+
+		if (monitor->enable_list_files) {
+			char *old_path = string_format("%s.files", output_prefix);
+			char *new_path = string_format("%s.files", log_prefix);
+			if (rename(old_path, new_path) == -1) {
+				debug(D_ERROR|D_MAKEFLOW_HOOK, "Error moving Resource Monitor output %s:%s. %s\n", old_path,
+						new_path, strerror(errno));
+				return MAKEFLOW_HOOK_FAILURE;
+			}
+			free(old_path);
+			free(new_path);
+		}
+
+		free(log_prefix);
+		free(output_prefix);
+	}
+	return MAKEFLOW_HOOK_SUCCESS;
+}
+
+static int node_end(struct dag_node *n, struct batch_task *task)
+{
+	char *log_prefix = set_log_prefix(n);
+	char *output_prefix = NULL;
+	if (batch_queue_supports_feature(makeflow_get_queue(n), "output_directories")) {
+		output_prefix = xxstrdup(log_prefix);
+	} else {
+		output_prefix = xxstrdup(path_basename(log_prefix));
+	}
+	char *summary_name = string_format("%s.summary", output_prefix);
+
+	if (n->resources_measured)
+		rmsummary_delete(n->resources_measured);
+	n->resources_measured = rmsummary_parse_file_single(summary_name);
+
+	/* If the resources_measured is null, then the expected files to move
+	 * are non-existent. This will cause the move function to fail and bail
+	 * on the makeflow as a whole, which we don't want. */
+	if(!n->resources_measured){
+		debug(D_MAKEFLOW_HOOK, "Resource Monitor failed to measure resources.\n");
+		return MAKEFLOW_HOOK_SUCCESS;
```

Comment:
Why we don't want to bail on this errror?

Reply:
Please correct me if this is wrong, but I was under the impression the only reason this would be null is if the summary file does not exist. If the resource monitor is enabled and the summary file doesn't exist, then the task failed before or while trying to run the resource monitor. It is also possible the file fails to transfer, but in any of these cases we shouldn't move non-existent files.

An alternative is to not error out if the files fail to transfer.

This speaks to a larger clarification on what is a fatal error for the resource monitor.

Reply:
I think I was confused. The task failed, but the hook did not. HOOK_SUCCESS does not mean task success, right? It simply means that the task was processed correctly by the hook.

Reply:
Yes. This is exactly correct. A failed node does not exit the makeflow, but a failed hook will.